subsys/bluetooth/host/gatt.c

```diff
@@ -278,13 +278,88 @@ static void sc_clear(struct gatt_sc_cfg *cfg)
 	memset(cfg, 0, sizeof(*cfg));
 }
 
+static void sc_reset(struct gatt_sc_cfg *cfg)
+{
+	BT_DBG("peer %s", bt_addr_le_str(&cfg->peer));
+
+	memset(&cfg->data, 0, sizeof(cfg->data));
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		sc_store(cfg);
+	}
+}
+
+static bool update_range(u16_t *start, u16_t *end, u16_t new_start,
+			 u16_t new_end)
+{
+	BT_DBG("start 0x%04x end 0x%04x new_start 0x%04x new_end 0x%04x",
+	       *start, *end, new_start, new_end);
+
+	/* Check if inside existing range */
+	if (new_start >= *start && new_end <= *end) {
+		return false;
+	}
+
+	/* Update range */
+	if (*start > new_start) {
+		*start = new_start;
+	}
+
+	if (*end < new_end) {
+		*end = new_end;
+	}
+
+	return true;
+}
+
+static void sc_save(u8_t id, bt_addr_le_t *peer, u16_t start, u16_t end)
+{
+	struct gatt_sc_cfg *cfg;
+	bool modified = false;
+
+	BT_DBG("peer %s start 0x%04x end 0x%04x", bt_addr_le_str(peer), start,
+	       end);
+
+	cfg = find_sc_cfg(id, peer);
+	if (!cfg) {
+		/* Find and initialize a free sc_cfg entry */
+		cfg = find_sc_cfg(BT_ID_DEFAULT, BT_ADDR_LE_ANY);
+		if (!cfg) {
+			BT_ERR("unable to save SC: no cfg left");
+			return;
+		}
+
+		cfg->id = id;
+		bt_addr_le_copy(&cfg->peer, peer);
+	}
+
+	/* Check if there is any change stored */
+	if (!(cfg->data.start || cfg->data.end)) {
+		cfg->data.start = start;
+		cfg->data.end = end;
+		modified = true;
+		goto done;
+	}
+
+	modified = update_range(&cfg->data.start, &cfg->data.end, start, end);
+
+done:
+	if (IS_ENABLED(CONFIG_BT_SETTINGS) &&
+	    modified && bt_addr_le_is_bonded(cfg->id, &cfg->peer)) {
+		sc_store(cfg);
+	}
+}
+
 static bool sc_ccc_cfg_write(struct bt_conn *conn,
 			     const struct bt_gatt_attr *attr,
 			     u16_t value)
 {
 	BT_DBG("value 0x%04x", value);
 
-	if (value != BT_GATT_CCC_INDICATE) {
+	if (value == BT_GATT_CCC_INDICATE) {
+		/* Create a new SC configuration entry if subscribed */
+		sc_save(conn->id, &conn->le.dst, 0, 0);
```

Comment:
Not following the reasoning here, so is this keeping the sc in memory for every paired device? Isn't that what we are trying to get rid of so we only consume ram for connected devices?

Reply:
When a database change occurs, we need to know which paired devices are subscribed to SC so we can save the SC data to be used next time the device reconnects. If we only have access to CCC data of connected devices, we need a way to know which devices are subscribed, so we can just keep an SC entry for that purpose.

It uses way less memory to only keep SC in memory rather than keeping data for all CCC attributes.